iso15118/shared/security.py

```diff
@@ -1117,6 +1117,69 @@ def decrypt_priv_key(
     raise DecryptionError()
 
 
+def derive_certificate_hash_data(certificate: bytes) -> Dict[str, str]:
+    """Extract certificate hash data to be used in an OCPP AuthorizeRequest.
+
+    Args:
+        certificate: The certificate in binary (DER) form.
+
+    Returns:
+        A dictionary with all information required for an OCSPRequestDataType
+        (2.36. OCSPRequestDataType, p. 382, OCPP 2.0.1 Part 2)
+
+    Raises:
+        CertAttributeError: if a certificate is provided with a hash algorithm
+            that OCPP doesn't accept.
+            Only SHA256, SHA384, and SHA512 are allowed.
+            (3.42 HashAlgorithmEnumType, p. 403, OCPP 2.0.1 Part 2)
+    """
+    certificate = load_der_x509_certificate(certificate)
+    public_key = certificate.public_key()
+    public_key_bytes = public_key.public_bytes(
+        encoding=Encoding.X962,
+        format=PublicFormat.UncompressedPoint,
+    )
+    # TODO: is this correct?
+    # gives a name like:
+    # 'DC=MO,C=DE,O=Keysight Technologies,CN=PKI-1_CRT_MO_SUB2_VALID'
+    # Do we need this whole thing or only, say, O?
+    distinguished_name = certificate.issuer.rfc4514_string().encode()
+    serial_number = certificate.serial_number
+
+    # Convert to the naming used in OCPP.
+    hash_algorithm_for_ocpp = certificate.signature_hash_algorithm.name.upper()
+    if hash_algorithm_for_ocpp not in {"SHA256", "SHA384", "SHA512"}:
+        raise CertAttributeError("Unknown hash algorithm")
+
+    public_key_hasher = Hash(certificate.signature_hash_algorithm)
+    public_key_hasher.update(public_key_bytes)
+    issuer_name_hasher = Hash(certificate.signature_hash_algorithm)
+    issuer_name_hasher.update(distinguished_name)
+
+    return {
+        "hash_algorithm": hash_algorithm_for_ocpp,
+        "issuer_name_hash": issuer_name_hasher.finalize(),
+        "issuer_key_hash": public_key_hasher.finalize(),
+        "serial_number": serial_number,
+        # TODO: Populate with a real-world verification server
+        "responder_url": "https://www.example.com/",
```

Comment:
I guess it will look something like this: "http://ocsp-qa.hubject.com:8080" (this is the QA env for the OCSP validation in Hubject.

Reply:
I did some research and this is what we need to do to extract the responder_url

```python
from cryptography.x509.oid import ExtensionOID, AuthorityInformationAccessOID

def get_ocsp_url(cert):
    auth_inf_access = cert.extensions.get_extension_for_oid(ExtensionOID.AUTHORITY_INFORMATION_ACCESS).value
    ocsps = [access_descriptor for access_descriptor in auth_inf_access if access_descriptor.access_method == AuthorityInformationAccessOID.OCSP]
    if not ocsps:
        raise Exception(f'no ocsp server entry in Authority Information Access extension field')
    return ocsps[0].access_location.value
```
If the Authority Information Access is not found, the method raises a `ExtensionNotFound` error

References: 
https://stackoverflow.com/questions/64436317/how-to-check-ocsp-client-certificate-revocation-using-python-requests-library

https://cryptography.io/en/latest/x509/reference/?highlight=AuthorityInformationAccessOID.CA_ISSUERS#cryptography.x509.Extensions.get_extension_for_oid

https://cryptography.io/en/latest/x509/reference/?highlight=AuthorityInformationAccessOID.OCSP#cryptography.x509.AccessDescription

https://cryptography.io/en/latest/x509/reference/?highlight=AuthorityInformationAccessOID.OCSP#cryptography.x509.oid.AuthorityInformationAccessOID

Reply:
Also, to get the hashed info for the OCSP request we can do this:
```python
from cryptography.x509.ocsp import OCSPRequestBuilder

def get_oscp_request(cert: bytes, issuer_cert: bytes):
    """
    cert: The certificate in DER format
    issuer_cert: The certificate used to sign the `cert` certificate, also called the issuer certificate
    """
    cert = load_der_x509_certificate(cert)
    issuer_cert = load_der_x509_certificate(issuer_cert)
    builder = OCSPRequestBuilder().add_certificate(cert, issuer_cert, cert.signature_hash_algorithm)
    return builder.build()

ocsp_request = get_oscp_request(certificate, issuer_cert)
```
The data can then be accessed through the returned object, for example
`ocsp_request.issuer_key_hash`.

Not sure if this is better as we need to load the issuer certificate, but makes it easier then to extract the needed data.

Reference:
https://cryptography.io/en/latest/x509/ocsp/#cryptography.x509.ocsp.OCSPRequest